### PR TITLE
[monorepo] feat: Add code coverage upload

### PR DESCRIPTION
**Pull request by Wayonb**
Add automatic code coverage upload to codecov.  PR will fail if the code coverage falls below the minimum configured value for the feature.

### CODE REVIEW

File: catbuffer/parser/scripts/ci/setup_lint.sh [hunk 1]
```diff
@@ -4,3 +4,4 @@ set -ex
 
 python3 -m pip install -r "$(git rev-parse --show-toplevel)/linters/python/lint_requirements.txt"
 python3 -m pip install -r requirements.txt
+python3 -m pip install -r dev_requirements.txt
```
Comment (Jaguar0625) [inline catbuffer/parser/scripts/ci/setup_lint.sh:7]: setup_lint or setup_test ?

File: catbuffer/parser/scripts/ci/test.sh [hunk 2]
```diff
@@ -2,4 +2,5 @@
 
 set -ex
 
-python3 -m unittest discover -v
+TEST_RUNNER=$([ "$1" = "code-coverage" ] && echo "coverage run" || echo "python3")
+${TEST_RUNNER} -m unittest discover -v
```
Comment (Jaguar0625) [inline catbuffer/parser/scripts/ci/test.sh:6]: nice job 👍

File: client/rest/package.json [hunk 3]
```diff
@@ -7,6 +7,7 @@
     "start": "node src/index.js",
     "lint": "eslint .",
     "test": "mocha --full-trace --recursive ./test",
+    "coverage:test": "nyc --require mocha --reporter=lcov npm run test",
     "lint:jenkins": "eslint -o lint.client.rest.xml -f junit . || exit 0",
     "test:jenkins": "JUNIT_REPORT_PATH=tests.client.rest.xml mocha --reporter mocha-jenkins-reporter --forbid-only --full-trace --recursive ./test || exit 0"
```
Comment (Jaguar0625) [inline client/rest/package.json:12]: do you still need test:jenkins ? if not, i'd replace it with coverage:test

File: codecov.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Setting coverage targets per flag
```
Comment (Jaguar0625) [inline codecov.yml:1]: will this file work with yaml extension? 😅

File: codecov.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+coverage:
+  status:
+    project:
+      default:
+        target: 80%  # overall project/ repo coverage
```
Comment (Jaguar0625) [inline codecov.yml:6]: spacing around '/'

File: codecov.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      sdk-javascript:
+        target: auto
+        flags:
+          - sdk-javascript
+
+# New root YAML section = `flags:`
+# This is where you would define every flag from your
+# uploader, and update when new Flags added
+
+flags:
```
Comment (Jaguar0625) [inline codecov.yml:32]: can you explain what this section does (different from previous)?

File: jenkins/docker/javascript.Dockerfile [hunk 7]
```diff
@@ -27,6 +27,11 @@ RUN apt-get install -y python3-pip
 RUN apt-get install -y shellcheck \
 	&& pip install gitlint
 
+# codecov uploader
+RUN curl -Os https://uploader.codecov.io/latest/linux/codecov \
+	&& chmod +x codecov \
+	&& mv codecov /usr/local/bin
```
Comment (Jaguar0625) [inline jenkins/docker/javascript.Dockerfile:33]: is this not in apt-get repo ?

File: jenkins/docker/python.Dockerfile [hunk 8]
```diff
@@ -18,7 +18,13 @@ RUN pip install poetry
...
 RUN apt-get install -y zbar-tools
 
+# codecov uploader
+RUN curl -Os https://uploader.codecov.io/latest/linux/codecov \
+	&& chmod +x codecov \
+	&& mv codecov /usr/local/bin
+
 # add ubuntu user (used by jenkins)
 RUN useradd --uid 1000 -ms /bin/bash ubuntu
+ENV PATH=$PATH:/home/ubuntu/.local/bin
```
Comment (Jaguar0625) [inline jenkins/docker/python.Dockerfile:28]: why is this needed in py docker but not js docker? 🤔

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 9]
```diff
@@ -23,13 +23,16 @@ void call(Closure body) {
 			choice name: 'BUILD_CONFIGURATION',
 				choices: ['release-private', 'release-public'],
 				description: 'build configuration'
+			choice name: 'TEST_MODE',
+					choices: ['code-coverage', 'test'],
+					description: 'test mode'
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy:28]: one indent too much

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 10]
```diff
@@ -23,13 +23,16 @@ void call(Closure body) {
 			choice name: 'BUILD_CONFIGURATION',
 				choices: ['release-private', 'release-public'],
 				description: 'build configuration'
+			choice name: 'TEST_MODE',
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy:26]: are you planning to run coverage nightly? or why are there code coverage and non - code coverage builds.

File: jenkins/shared-library/vars/runScript.groovy [hunk 11]
```diff
@@ -6,7 +6,7 @@ void call(String scriptPath, String filename) {
 }
 
 void call(String scriptFilepath) {
-	call(scriptFilepath, scriptFilepath, true)
+	call(scriptFilepath, scriptFilepath, false)
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/runScript.groovy:9]: why this change?

File: sdk/javascript/.nycrc.yaml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+exclude:
+        - '**/external/**'
```
Comment (Jaguar0625) [inline sdk/javascript/.nycrc.yaml:2]: indented too much

File: sdk/javascript/package.json [hunk 13]
```diff
@@ -6,7 +6,9 @@
   "scripts": {
     "lint": "eslint .",
     "test": "mocha --full-trace --recursive ./test",
+    "coverage:test": "nyc --require mocha --reporter=lcov npm run test",
     "vectors": "node ./vectors/all.js --vectors $(git rev-parse --show-toplevel)/tests/vectors/${BLOCKCHAIN}/crypto --blockchain ${BLOCKCHAIN}",
+    "coverage:vectors": "nyc --require mocha --no-clean --reporter=lcov npm run vectors",
```
Comment (Jaguar0625) [inline sdk/javascript/package.json:11]: if this only for jenkins, can rename to vectors:jenkins

File: sdk/python/.coveragerc [hunk 14]
```diff
@@ -0,0 +1,2 @@
+[run]
+omit = **/dist-packages/**
```
Comment (Jaguar0625) [inline sdk/python/.coveragerc:2]: also should omit externals

File: sdk/python/scripts/ci/test_vectors.sh [hunk 15]
```diff
@@ -2,7 +2,9 @@
 
 set -ex
 
-PYTHONPATH=. SCHEMAS_PATH="$(git rev-parse --show-toplevel)/tests/vectors" python3 -m pytest tests/generator
+TEST_RUNNER=$([ "$1" = "code-coverage" ] && echo "coverage run --append" || echo "python3")
```
Comment (Jaguar0625) [inline sdk/python/scripts/ci/test_vectors.sh:5]: is it intentional that this is only command with `--append` ?

File: catbuffer/parser/scripts/ci/setup_lint.sh (hunk 1 above)
Comment (Wayonb) [reply catbuffer/parser/scripts/ci/setup_lint.sh:7]: yea, moved to setup_test better

File: catbuffer/parser/scripts/ci/test.sh (hunk 2 above)
Comment (Wayonb) [reply catbuffer/parser/scripts/ci/test.sh:6]: Thanks

File: client/rest/package.json (hunk 3 above)
Comment (Wayonb) [reply client/rest/package.json:12]: Yea, can remove test:jenkins

File: codecov.yml (hunk 4 above)
Comment (Wayonb) [reply codecov.yml:1]: let me see.

File: codecov.yml (hunk 6 above)
Comment (Wayonb) [reply codecov.yml:32]: I can add some more comments to it.  The flags section actually define the flags(root folder for the package) that are used in the projects above.

File: jenkins/docker/javascript.Dockerfile (hunk 7 above)
Comment (Wayonb) [reply jenkins/docker/javascript.Dockerfile:33]: No, they had an older version I think in apt.
https://docs.codecov.com/docs/codecov-uploader

File: jenkins/docker/python.Dockerfile (hunk 8 above)
Comment (Wayonb) [reply jenkins/docker/python.Dockerfile:28]: I saw some warnings when installing the python packages

File: jenkins/shared-library/vars/defaultCiPipeline.groovy (hunk 9 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/defaultCiPipeline.groovy:28]: ahh...got to stick with one editor

File: jenkins/shared-library/vars/defaultCiPipeline.groovy (hunk 10 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/defaultCiPipeline.groovy:26]: No, was thinking for Release builds probably dont need to run code coverage tests.   Plus if there is an issue with tests can run them without.  Can always leave it enabled if you want

File: jenkins/shared-library/vars/runScript.groovy (hunk 11 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/runScript.groovy:9]: some logging was missing in Jenkins logs.

File: sdk/javascript/package.json (hunk 13 above)
Comment (Wayonb) [reply sdk/javascript/package.json:11]: would run locally too so prefer to remove the `` vectors:jenkins``

Comment (Wayonb) [reply sdk/javascript/package.json:11]: Thinking about this more would only keep the jenkins scripts if we still plan to have nightly runs.

File: sdk/python/scripts/ci/test_vectors.sh (hunk 15 above)
Comment (Wayonb) [reply sdk/python/scripts/ci/test_vectors.sh:5]: Yes, by default code coverage tool will overwrite the previous run.  
The other option which I was also thinking would be to generate a different code coverage file for each run and merged them  at the end.

File: client/rest/package.json (hunk 3 above)
Comment (Jaguar0625) [reply client/rest/package.json:12]: i'd replace test:jenkins with coverage:test in that case

File: jenkins/docker/python.Dockerfile (hunk 8 above)
Comment (Jaguar0625) [reply jenkins/docker/python.Dockerfile:28]: @gimre-xymcity - any idea?

File: sdk/javascript/package.json (hunk 13 above)
Comment (Jaguar0625) [reply sdk/javascript/package.json:11]: you want nightlies without code coverage?

File: sdk/python/scripts/ci/test_vectors.sh (hunk 15 above)
Comment (Jaguar0625) [reply sdk/python/scripts/ci/test_vectors.sh:5]: shouldn't js test-vectors also have append or does tool automatically merge?

File: client/rest/package.json (hunk 3 above)
Comment (Wayonb) [reply client/rest/package.json:12]: ok

File: sdk/javascript/package.json (hunk 13 above)
Comment (Wayonb) [reply sdk/javascript/package.json:11]: We can run code coverage at night also but it should be the same as last check-in to dev I think.  When we start running e2e then code coverage should be enabled since those dont run on every check-in.  Nightly would be more to collect test results.

File: sdk/python/scripts/ci/test_vectors.sh (hunk 15 above)
Comment (Wayonb) [reply sdk/python/scripts/ci/test_vectors.sh:5]: its part of the script, there is a ``--no-clean`` flag.  
I am think I should probably add the ``append`` flag to all of the code coverage run so order would matter.  The first run will should always be in a clean workspace so it shouldnt matter.  Your thoughts?

File: client/rest/package.json [hunk 16]
```diff
@@ -7,9 +7,8 @@
     "start": "node src/index.js",
     "lint": "eslint .",
     "test": "mocha --full-trace --recursive ./test",
-    "coverage:test": "nyc --require mocha --reporter=lcov npm run test",
-    "lint:jenkins": "eslint -o lint.client.rest.xml -f junit . || exit 0",
-    "test:jenkins": "JUNIT_REPORT_PATH=tests.client.rest.xml mocha --reporter mocha-jenkins-reporter --forbid-only --full-trace --recursive ./test || exit 0"
+    "test:jenkins": "nyc --require mocha --reporter=lcov npm run test",
+    "lint:jenkins": "eslint -o lint.client.rest.xml -f junit . || exit 0"
```
Comment (Jaguar0625) [inline client/rest/package.json:11]: would prefer you keep ordering
lint
test
lint:jenkins
test:jenkins

File: sdk/javascript/scripts/ci/test.sh [hunk 17]
```diff
@@ -2,5 +2,5 @@
 
 set -ex
 
-TEST_MODE=$([ "$1" = "code-coverage" ] && echo "coverage:test" || echo "test")
-npm run "${TEST_MODE}"
+TEST_MODE=$([ "$1" = "code-coverage" ] && echo "test:jenkins" || echo "test")
+TEST_SCRIPT='test' npm run "${TEST_MODE}"
```
Comment (Jaguar0625) [inline sdk/javascript/scripts/ci/test.sh:6]: this isn't running anything is it?

File: sdk/javascript/scripts/ci/test_vectors.sh [hunk 18]
```diff
@@ -2,6 +2,7 @@
 
 set -ex
 
-TEST_MODE=$([ "$1" = "code-coverage" ] && echo "coverage:vectors" || echo "vectors")
+TEST_MODE=$([ "$1" = "code-coverage" ] && echo "test:jenkins" || echo "vectors")
+export TEST_SCRIPT='vectors'
```
Comment (Jaguar0625) [inline sdk/javascript/scripts/ci/test_vectors.sh:6]: why this ?

File: sdk/javascript/scripts/ci/test.sh (hunk 17 above)
Comment (Wayonb) [reply sdk/javascript/scripts/ci/test.sh:6]: Instead of having two scripts for code coverage(test and vectors), I created a env variable which state which to run.  It should run either test or vectors.
I can move back to two scripts if that is more readable.

File: client/rest/package.json (hunk 16 above)
Comment (Wayonb) [reply client/rest/package.json:11]: ok

File: sdk/javascript/scripts/ci/test_vectors.sh (hunk 18 above)
Comment (Wayonb) [reply sdk/javascript/scripts/ci/test_vectors.sh:6]: explained above.

File: sdk/javascript/scripts/ci/test.sh (hunk 17 above)
Comment (Jaguar0625) [reply sdk/javascript/scripts/ci/test.sh:6]: i think two would be more readable

File: client/rest/package.json (hunk 3 above)
Comment (Jaguar0625) [reply client/rest/package.json]: can you just reorder this one too (lint, test, lint:jenkins, test:jenkins)

File: jenkins/docker/python.Dockerfile (hunk 8 above)
Comment (gimre-xymcity) [reply jenkins/docker/python.Dockerfile:28]: hmm not really, do you perhaps have error saved?

Comment (Wayonb) [reply jenkins/docker/python.Dockerfile:28]: @gimre-xymcity This was the warning I saw in the logs

```
13:47:33  + python3 -m pip install -r /home/ubuntu/jenkins/workspace/l_generated_monorepo_parser_main/linters/python/lint_requirements.txt
13:47:35    WARNING: The scripts isort and isort-identify-imports are installed in '/home/ubuntu/.local/bin' which is not on PATH.
13:47:35    Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
```

**Review by Jaguar0625 — APPROVED** [review]
